### DISCUSSION TITLE
Highlighting php without "<?php"

### DISCUSSION THREAD

**Discussion by fabiospampinato** [github-linguist/linguist#5553] [Q&A] [answered]
Hello there, I'm trying to have syntax highlighting for all the languages that github itself supports using the same grammars too.

I'm hitting a problem that I'm not sure how to solve at the moment, PHP code blocks containing something like the following (e.g. without "<?php") are highlighted by GitHub:

```php
// Hello world in PHP
echo 'Hello World!';
```

But they don't seem to be supported by the underlying [TextMate grammar](https://github.com/textmate/php.tmbundle/blob/8f05764681d51cd0b13db1e36810d74913c86bfc/Syntaxes/PHP.plist):

<img width="385" alt="Screen Shot 2021-09-03 at 19 21 58" src="https://user-images.githubusercontent.com/1812093/132045068-6cfdd0f3-29bf-46fc-a5b5-899e90660fc4.png">

According to a [comment](https://github.com/github/linguist/issues/4865#issuecomment-630626614) from @maxbrunsfeld it seems like some languages are special-cased, presumably to fix issues such as the one I'm getting.

If so is that code open source by any chance? If not does anybody have any clues on how to fix this?

Also is there a list somewhere of all the languages that are special-cased by github for syntax highlighting purposes?

**Comment by lildude** [comment, marked as answer]
> If so is that code open source by any chance?

No. It's within the main GitHub.com code which isn't open source. It essentially detects the scope `text.html.php` and prepends `<?php\n` to the text if it's not there, colourises it, and then strips it from the beginning of the output before rendering. This only applies to inline code snippets.

>  If not does anybody have any clues on how to fix this?

Nope, but I know VSCode has the same behaviour:

![CleanShot 2021-09-06 at 14 37 49](https://user-images.githubusercontent.com/627280/132225834-d86b7100-b5ad-4c4e-b8d6-41fb6bd508f4.png)

![CleanShot 2021-09-06 at 14 38 12](https://user-images.githubusercontent.com/627280/132225849-8661e8ab-fbf5-43d4-b9c5-8559ee7c5972.png)

That said, you're not really comparing like-for-like.  Inline snippets are code blocks in Markdown, so if you change your TextMate example to use Markdown, I'd expect you to get the same behaviour. I've not tested it as I don't have TextMate installed, but this is how VSCode works:

![CleanShot 2021-09-06 at 14 50 13](https://user-images.githubusercontent.com/627280/132227351-01b67edf-8ecd-4a10-94b5-ed84a6b3a0af.png)

> Also is there a list somewhere of all the languages that are special-cased by github for syntax highlighting purposes?

PHP is the only one.

>  I'm trying to have syntax highlighting for all the languages that github itself supports using the same grammars too.

As an FYI: We convert all grammars into JSON and attach a tarball of all of these files to each release if you want to use grammars exactly as they're used by GitHub. One thing to keep in mind, is we use the public [tree-sitter](https://github.com/tree-sitter) grammars for the following languages and not those included in Linguist:

- CSS
- Go
- JavaScript
- EJS
- ERB
- HTML
- Python
- Ruby
- TypeScript
- CodeQL

We still bundle these grammars with Linguist so other grammars can still pull in components of each and also for those languages that share a common grammar but with different language names.

**Reply by fabiospampinato** [reply]
Thanks for the treasure of information!

For the interest of future readers [this grammar](https://github.com/shikijs/shiki/blob/54fb99b4be1a59059c24739ff7fe05f03ae8a26d/packages/shiki/languages/php.tmLanguage.json) simply supports highlighting PHP that doesn't have the `<?php` prefix too.
